packages/@blockprotocol/core/src/core-handler.ts

```diff
@@ -0,0 +1,345 @@
+  protected element: HTMLElement;
+
+  /** whether the instance of CoreHandler belongs to a block or embedding application */
+  protected readonly sourceType: "block" | "embedder";
+
+  /** when this handler itself sends messages, serviceName identifies it as the core handler */
+  readonly serviceName: "core" = "core";
+
+  /**
+   * Handles the initialization messages exchanged when a block registers a handler
+   * 1. the block sends a message
+   * 2. the embedder calls processInitMessage to generate a response,
+   *   including initial values for messages which are 'sentOnInitialization: true'
+   * 3. the block calls processInitMessage to process the response and values sent in (2)
+   * */
+  protected abstract processInitMessage(args: {
+    event: CustomEvent;
+    message: Message;
+  }): void;
+
+  /** the event name used to transport messages */
+  private static readonly customEventName = "blockprotocolmessage";
+
+  /**
+   * A map of instances of CoreHandler, by the element they are listening on,
+   * used to ensure that two handlers aren't instantiated for the same element.
+   */
+  private static readonly instanceMap = new Map<HTMLElement, CoreHandler>();
+
+  private static isBlockProtocolMessage(message: unknown): message is Message {
+    if (
+      typeof message !== "object" ||
+      message === null ||
+      Array.isArray(message)
+    ) {
+      return false;
+    }
+    if (
+      !("requestId" in message) ||
+      !("service" in message) ||
+      !("source" in message) ||
+      !("messageName" in message)
+    ) {
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Register a ServiceHandler for the given element.
+   * Will re-use the CoreHandler if it exists for that element, or create a new one.
+   * Currently the only way CoreHandlers may be instantiated
+   *
+   * [FUTURE]: there may be a need to instantiate CoreHandlers without any services
+   */
+  static registerService({
+    element,
+    service,
+  }: {
+    element: HTMLElement;
+    service: ServiceHandler;
+  }) {
+    const { serviceName } = service;
+    const handler =
+      this.instanceMap.get(element) ?? Reflect.construct(this, [{ element }]);
+    handler.services.set(serviceName, service);
+    handler.messageCallbacksByService[serviceName] ??= new Map();
+    return handler;
+  }
+
+  unregisterService({ service }: { service: ServiceHandler }) {
+    const { serviceName } = service;
+    this.services.delete(serviceName);
+    if (this.services.size === 0) {
+      this.removeEventListeners();
+      CoreHandler.instanceMap.delete(this.element);
+    }
+  }
+
+  protected constructor({
+    element,
+    sourceType,
+  }: {
+    element: HTMLElement;
+    sourceType: "block" | "embedder";
+  }) {
+    this.responseSettlersByRequestIdMap = new Map();
+    this.messageCallbacksByService = {};
+    this.services = new Map();
```

Comment:
If you mean why not defaults where declared, no particular reason. Do you think setting as defaults is better?

If you mean why not leave them undefined until being used, I think it's less hassle to have them guaranteed to exist when the instance is created, and if the handler is going to be used at all they'll all be necessary (rather than being able to leave some undefined indefinitely).

Reply:
> Do you think setting as defaults is better?

Yeah, setting where the param is defined vs setting in constructor is functionally the same, but the first is cleaner

Reply:
https://github.com/blockprotocol/blockprotocol/pull/323/commits/6a1ee261c702e60df6d0493c8faf3e2c2101b2cd

Reply:
👍